src/features/core/metrics/errorReporting/providers/Bugsnag.js

```diff
@@ -0,0 +1,233 @@
+const BUGSNUG_LIB_INSTANCE_GLOBAL_KEY_NAME = 'bugsnag'; // For version 7 this is 'Bugsnug'
+const GLOBAL_LIBRARY_OBJECT_NAMES = [
+  ERROR_REPORT_PROVIDER_NAME_BUGSNUG,
+  'Bugsnug',
+  BUGSNUG_LIB_INSTANCE_GLOBAL_KEY_NAME,
+  ERROR_REPORTING_SERVICE_GLOBAL_KEY_NAME,
+];
+
+// This API key token is parsed in the CI pipeline
+const API_KEY = '__RS_BUGSNAG_API_KEY__';
+
+// Errors only from Below SDKs are allowed to reach Bugsnag
+const SDK_FILE_NAMES = [
+  'rudder-analytics.min.js',
+  'rudder-analytics-staging.min.js',
+  'rudder-analytics.js',
+  ...Object.keys(configToIntNames).map((intgName) => `${configToIntNames[intgName]}.min.js`),
+  ...Object.keys(configToIntNames).map(
+    (intgName) => `${configToIntNames[intgName]}-staging.min.js`,
+  ),
+  ...Object.keys(configToIntNames).map((intgName) => `${configToIntNames[intgName]}.js`),
+];
+const BUGSNAG_ERROR_FILE_ORIGIN_KEY = 'errors.0.stacktrace.0.file';
+
+const getReleaseStage = () => {
+  const host = window.location.hostname;
+  const devHosts = ['www.rs-unit-test-host.com', 'localhost', '127.0.0.1', '[::1]'];
+
+  return host && devHosts.includes(host) ? 'development' : '__RS_BUGSNUG_RELEASE_STAGE__';
+};
+
+const isValidVersion = (globalLibInstance) => {
+  // For version 7
+  let version =
+    globalLibInstance &&
+    globalLibInstance._client &&
+    globalLibInstance._client._notifier &&
+    globalLibInstance._client._notifier.version;
+
+  // For versions older than 7
+  if (!version) {
+    let tempInstance = globalLibInstance({ apiKey: API_KEY, releaseStage: 'version-test' });
+    version = tempInstance.notifier && tempInstance.notifier.version;
+    tempInstance = undefined;
+  }
+
+  return version && version.charAt(0) === BUGSNUG_VALID_MAJOR_VERSION;
+};
+
+const isRudderSDKError = (event) => {
+  const errorOrigin = get(event, BUGSNAG_ERROR_FILE_ORIGIN_KEY);
+
+  if (!errorOrigin || typeof errorOrigin !== 'string') {
+    return false;
+  }
+
+  const srcFileName = errorOrigin.substring(errorOrigin.lastIndexOf('/') + 1);
+  return SDK_FILE_NAMES.includes(srcFileName);
+};
+
+const enhanceErrorEventMutator = (event, metadataSource) => {
+  event.addMetadata('source', {
+    metadataSource,
+  });
+
+  const errorMessage = event.errors && event.errors[0] && event.errors[0].errorMessage;
+  // eslint-disable-next-line no-param-reassign
+  event.context = errorMessage;
+
+  // Hack for easily grouping the script load errors
+  // on the dashboard
+  if (errorMessage.includes('error in script loading')) {
+    // eslint-disable-next-line no-param-reassign
+    event.context = 'Script load failures';
+  }
+
+  // eslint-disable-next-line no-param-reassign
+  event.severity = 'error';
+};
+
+const loadBugsnugSDKScript = (name) => {
+  const isNotLoaded = GLOBAL_LIBRARY_OBJECT_NAMES.every(
+    (globalKeyName) => !Object.hasOwn(window, globalKeyName),
+  );
+
+  if (isNotLoaded) {
+    ScriptLoader(name, BUGSNUG_CDN_URL, {
```

Comment:
Is it possible that this throws an error? if yes, then should we catch this error?

Reply:
you mean the script loader function? This is used now as is for integrations

Reply:
Yes scriptloader, if this throws error then it won't be caught.

Reply:
Will put the try catch inside the ScriptLoader so it is not blowing up when used in other places too